src/shared/components/RichTable.tsx

```diff
@@ -0,0 +1,666 @@
+      if (searchQuery) {
+        queryBuilder.search = window.encodeURIComponent(searchQuery);
+      }
+      if (sortColumn !== null) {
+        queryBuilder.sort = (sortColumn[1] ? "" : "-") + window.encodeURIComponent(sortColumn[0]);
+      }
+      return `${url}${Object.keys(queryBuilder).length ? "?" : ""}${Object.entries(queryBuilder)
+        .map(q => q.join("="))
+        .join("&")}`;
+    },
+    // We can ignore the warning about cuckooLoad being an unnecessary dependency, since it exists to trigger
+    // refreshing without changing other state.
+    [url, paginated, sortColumn, searchQuery, page, cuckooLoad]
+  );
+  const [rawData, error] = useApiGet<Pagination<D> | D[]>(dataUrl);
+  const data = useMemo<D[]>(
+    () =>
+      (paginated ? (rawData as Pagination<D>)?.results : (rawData as D[])) ||
+      [],
+    [paginated, rawData]
+  );
+
+  const numPages = paginated ? (rawData as Pagination<D>)?.num_pages || 0 : 0;
+  const totalCount = paginated
+    ? (rawData as Pagination<D>)?.count || 0
+    : data.length;
+
+  const selectAllRef = useRef<HTMLInputElement & FormCheck>(null);
+  const [selected, setSelected] = useState<boolean[]>([]);
+  const [filteredSelected, setFilteredSelected] = useState<D[]>([]);
+
+  useEffect(() => {
+    setSelected(new Array(data.length).fill(false));
+  }, [data]);
+  useEffect(() => {
+    if (selectAllRef.current) {
+      selectAllRef.current.indeterminate =
+        !selected.every(d => d) && selected.some(d => d);
+    }
+  }, [selected]);
+  useEffect(() => {
+    setFilteredSelected(data.filter((_, i) => selected[i]));
+  }, [data, selected]);
+
+  const clickActions = useMemo<Action<D>[]>(
+    () =>
+      actions.filter(
+        action => action.triggers && action.triggers.includes("click")
+      ),
+    [actions]
+  );
+
+  const memoizedActions = useMemo<{ [key: string]: Action<D> }>(
+    () => ({
+      refresh: {
+        name: "refresh",
+        displayName: "Refresh",
+        call() {
+          setCuckoo(cuckoo => !cuckoo);
+          return Promise.resolve();
+        }
+      },
+      ...actions.reduce(
+        (acc, action) => ({ ...acc, [action.name]: action }),
+        {}
+      )
+    }),
+    [actions]
+  );
+
+  const internalExecuteAction = useCallback(
+    (name: string, rows: D[]) => {
+      const action = memoizedActions[name];
+
+      if ("bulk" in action) {
+        if (action.bulk) {
+          return action.call(bag, rows);
+        } else {
+          if (rows.length === 1) {
+            return action.call(bag, rows[0]);
+          } else {
+            return Promise.reject(
+              "Multiple rows selected for non-bulk action."
+            );
+          }
+        }
+      } else {
+        return action.call(bag);
+      }
+    },
+    // We can ignore the warning about bag being a missing dependency -- although this technically depends on
+    // it, both depend on data and so will refresh appropriately.
+    [data, memoizedActions]
```

Comment:
To be frank, I'm not quite comfortable with going against the linter on this. While I think we should strive to get rid of all these cases, at the very least, we should drop an explicit `// eslint-disable-next-line react-hooks/exhaustive-deps` so we don't get used to having a bunch of warnings in the dev server, and if anything breaks, we can go back and look there first.

Reply:
There's a big list of common "linter complaining at me scenarios" [here](https://github.com/facebook/react/issues/14920#issuecomment-471070149), if that helps any.

Reply:
I should note that because of how JS does block-level variable resolution, I can't put `bag` in the dependencies. My worry about ignoring exhaustive-deps here was that it might cover an actual problem later on; I'd rather have a single warning in the dev server for which we have documentation that we can ignore, rather than missing out on legitimate warnings later.

Reply:
On the other hand, if warnings pop up in other parts of the code, they might get lost in the middle of the "warnings we're OK with". Not to mention we'll have to get CI to ignore warnings as well.

Reply:
We can go on an eslint-hunting run later on; that will cut down on the number of warnings that show up, especially since a lot of them have to be silenced. Once we get to that point, we'll be left with only that warning, hopefully.

Reply:
Fair enough, we've probably spent enough time bikeshedding on this as it is. By the power vested in me by the Province of Ontario, I hereby resolve this conversation.